### ISSUE TITLE
Make operation attributes public

### ISSUE THREAD

**Issue by lancenonce** [sonos/tract#1214]
#### Problem
When using the tract library, attribute data for operations are typically private. This makes developing applications that need this data more difficult. 

#### Example
In tract/hir/src/ops/nn/layer_max.rs, `LogSoftmax` is implemented with no getters for `axis` or `coerce_to_2d`:
```
#[derive(Debug, Clone, new, Default, Hash)]
pub struct LayerLogSoftmax {
    axis: isize,
    coerce_to_2d: bool,
}
```
In my project that uses LogSoftmax, I cannot access `axis`

#### Solution
Add getters for op attributes so developers can do more with the tract library.

**Comment by kali** [comment]
Feel free to submit PR to the attributes that you need. I will not commit on stability, but they are super unlikely to change, so why not.

Depending on what is your end-game, you may want to link against the TypedOp core Op set instead of the hir and onnx ones (and declutter your model before anything else). A big part of what tract "decluttering" does is reduce the wide training-aware operator sets that onnx and tensorflow (and even tflite to some extend) to this smaller one.

**Comment by lancenonce** [comment]
Thanks for the prompt response! The scope of the project is to collect and utilize ONNX data. We don't perform any inference, we just collect operations, parameters (weights and biases), and attributes for later use. 

We parse incoming models as InferenceModels because we want to have access to as much information as possible (doesn't matter if it comes from onnx, hir, or core). We do, however, optimize and collect attributes on TypedOps such as Softmax. This implementation was straightforward in core. However, I didn't see an implementation of LogSoftmax in core, only in hir. My assumption here was that maybe this was only processed in hir and, consequently, that data is only available in hir? That was why I requested getters for those attributes. 

For reference, here are the operations that aren't in core we need attribute access to (will be more in the future):
- ArgMax(bool) (is the bool for keepdims?)
- ArgMin(bool) (same question)
- cumsum
- DequantizeLinearf32 (in core, just need getters since attributes aren't public)
- QuantizeLinearu8 (same, just need getters)
- Flatten
- ReduceSum
- Reshape (or Tensor::set_shape() if that's detected in the model)
- Transpose
- LogSoftmax

I'd be more than happy to implement getters for these and any other operations that would be relevant. Just let me know if I'm wrong about my prior assumption regarding some of these only being available in hir.

**Comment by kali** [comment]
LogSoftMax, CumSum, Flatten at least are defined as Expansion, so tract "decluttering" will expand them on combination of operators implemented with the smaller operator set in tract-core. (I have not double checked all of them and don't trust my memory).

The bool in ArgMax and ArgMin model onnx's "select_last_index" attribute. See https://github.com/sonos/tract/blob/main/onnx/src/ops/nn/mod.rs#L16 .

If you do a PR, don't do getters, just make the attributes public, we don't want the extra complexity, this stuff is pretty stable.

**Comment by lancenonce** [comment]
Is there any way we can take this conversation to Twitter? Would help a lot for future problems (sometimes an issue isn't the best plan of action). Here is my profile: https://twitter.com/lancenonce

I'm afraid I don't see an implementation for `declutter` on LogSoftmax since it doesn't implement TypedOp. Would love to have a chat about this.

And thanks for the clarity on ArgMax and ArgMin. I will start a fork with the new implementations and verify that the project works before submitting the PR.

**Comment by kali** [comment]
Sorry, I mispoke before. Expansion (logsoftmax and the like) are like macros: the wire() function translates them to a network of several TypedOp.

I'm slowly trying to stop using twitter, so I prefer to keep things on GH, in issues or "discussions" if it is not directly actionable.

**Comment by lancenonce** [comment]
I see! so calling Logsoftmax::wire will produce a TypedOp with public attributes?

And that's fine! thanks for the clarity

**Comment by kali** [comment]
Logsoftmax::wire will build several ops inside the mutable model argument and returns the outputs identifiers of the LogSoftmax. 

In the end wire() should creates a tract core SoftMax chained with a Ln.

See:
https://github.com/sonos/tract/blob/main/hir/src/ops/nn/layer_max.rs#L103
https://github.com/sonos/tract/blob/main/hir/src/ops/nn/layer_max.rs#L137

**Comment by lancenonce** [comment]
Ah, now I see. In this case, I think the PR would be relevant then. It makes sense for computation to do this, but for my project I just need access to the inner attributes of the hir representation. If we use `wire()` to convert to Softmax, I'm not sure if the operation would still register as LogSoftmax.

Am I good to submit a PR making the needed attributes public?

**Comment by kali** [comment]
OK. All good then, make public the attributes you need.

**Comment by lancenonce** [comment]
Thank you! Just a couple more questions:

- Do you know what the best way to get the target shape from Reshape would be? https://github.com/sonos/tract/blob/ff5657ec1144f637ef1c294f4c394654e4ce60d6/hir/src/ops/array/reshape.rs#L37 Would like to have access to the `shape` value here, but not sure how to retrieve it
- Same question for ArgMax/Min, the Reducer implementation is clever and convenient, but not sure how to retrieve the intermediary values: https://github.com/sonos/tract/blob/main/onnx/src/ops/nn/mod.rs#L20
